### DISCUSSION TITLE
An official Docker image to run quarto

### DISCUSSION THREAD

**Discussion by gdurif** [quarto-dev/quarto-cli#3374] [Feature Requests]
Hi,

I use quarto to build a static website that I host through gitlab pages (and automatically build through gitlab CI).

To do so, gitlab relies on Docker images from Docker Hub, so I had to find an image running smoothly quarto. I am currently configuring a Docker image so quarto runs smoothly within it (including rendering for any Python/R/Julia code chunk).

Having an official quarto images would be great (or multiple ones, e.g. quarto-R, quarto-python, quarto-julia, quarto-all, etc.).

Thanks in advance,
Best

**Comment by mcanouil** [comment]
There is the rocker project: <https://rocker-project.org/>
Also @rgaiacs did some work on the subject: <https://gitlab.com/quarto-forge/docker>.

**Comment by gdurif** [comment]
Hi, thanks for the suggestions.

The rocker project is great but R centered. I tried to use the [`rocker/rstudio`](https://github.com/rocker-org/rocker-versioned2/blob/master/dockerfiles/rstudio_devel.Dockerfile) image which contains quarto but it did not work when having both R and python code chunks within my markdown files.

Regarding the second link, the image `registry.gitlab.com/quarto-forge/docker/polyglot` is just what I need. I did not know that it was possible to use docker images not hosted on Docker hub, it is great!

**Reply by mcanouil** [reply]
Not everything is installed in the Rocker images, but the scripts are embedded in it and for some, they can be executed at run time using Docker options/parameters.

To note, GitHub has also its own Docker registry.

**Reply by rgaiacs** [reply]
@gdurif I'm glad that `registry.gitlab.com/quarto-forge/docker/polyglot` is what you were looking for. The Docker image is a little bigger at the moment because it include some tools to convert Quarto to PDF. I'm planning to refactor things a little in January/2023 to have a smaller image limited to HTML and a bigger one for PDF.

**Comment by eitsupi** [comment]
In general, installation of quarto-cli is easy, so you can simply install quarto-cli on any Docker image.

**Comment by mcanouil** [comment]
Closing as there is now Quarto docker images built within the repository in addition to many more elsewhere, see <https://github.com/quarto-dev/quarto-cli/pkgs/container/quarto>.

**Comment by jcrben** [comment]
@mcanouil the unfortunate thing about the images is that they don't contain `knitr` and `rmarkdown` - I'm trying to do the VSCode Quarto Preview on Universal Blue Aurora and haven't had success so far - had to install `brew install gcc-11`, then ultimately hit this after doing a bunch of compilation:

```
** testing if installed package can be loaded from temporary location
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (knitr)
ERROR: dependency ‘sass’ is not available for package ‘bslib’
* removing ‘/var/home/linuxbrew/.linuxbrew/lib/R/4.4/site-library/bslib’
ERROR: dependency ‘bslib’ is not available for package ‘rmarkdown’
* removing ‘/var/home/linuxbrew/.linuxbrew/lib/R/4.4/site-library/rmarkdown’

The downloaded source packages are in
	‘/tmp/RtmpJHkvgn/downloaded_packages’
Warning messages:
1: In install.packages(c("rmarkdown", "knitr")) :
  installation of package ‘fs’ had non-zero exit status
2: In install.packages(c("rmarkdown", "knitr")) :
  installation of package ‘sass’ had non-zero exit status
3: In install.packages(c("rmarkdown", "knitr")) :
  installation of package ‘bslib’ had non-zero exit status
4: In install.packages(c("rmarkdown", "knitr")) :
  installation of package ‘rmarkdown’ had non-zero exit status
> 
``` 

I installed RStudio in a "BoxBuddy" which works OK to see what Quarto can do, but ultimately I'm looking to use VSCode. 

I'll admit this Universal Blue thing is probably more complex than most, I'm still figuring it out.

**Reply by jcrben** [reply]
even my boxbuddy instance running debian 12 isn't going to well...
![image](https://github.com/user-attachments/assets/5617082e-ec7c-4be6-b504-3893c76c827a)

Clicked play on my Fedora 36 one and it compiled a bunch of stuff before finally failing on some knitr thing

**Reply by mcanouil** [reply]
Computing dependencies are unrelated to Quarto CLI, as those evolve quite quickly that's not possible for the Quarto team to maintain this, possibly for multiple distribution.

If you want to use less common distribution for whatever reasons, you need to install the R packages dependencies before installing the R packages. Those are listed by the package maintainers on the CRAN page for instance.

I personally have a Docker image with Julia, Python, and R, in their latest version that I have made which uses Ubuntu Jammy.

**Reply by jcrben** [reply]
I ended up being able to render this `quarto` executable file:

```
#!/usr/bin/env bash
# Translate host paths to container paths
ARGS=("$@")
for i in "${!ARGS[@]}"; do
  if [[ "${ARGS[$i]}" == /* ]]; then
    ARGS[$i]="/host${ARGS[$i]}"
  fi
done
  # -v /home/ben:/home/ben \
docker run --rm \
  -v /:/host \
  -v /home/linuxbrew/.linuxbrew:/home/linuxbrew/.linuxbrew \
  -e PATH="/home/linuxbrew/.linuxbrew/bin:/home/ben/.local/bin:$PATH" \
  -w /host$(pwd) \
  --network="host" \
  myimage2 quarto "${ARGS[@]}"
```

where `myimage2` is built with this:

```
FROM rocker/r-ver:4.4.0

RUN apt-get update && apt-get install -y \
    wget \
    python3 \
    python3-pip \
    python3-venv

# Install Quarto
RUN wget https://github.com/quarto-dev/quarto-cli/releases/download/v1.5.56/quarto-1.5.56-linux-amd64.deb && \
    dpkg -i quarto-1.5.56-linux-amd64.deb && \
    rm quarto-1.5.56-linux-amd64.deb

RUN python3 -m pip install --upgrade pip && \
python3 -m pip install jupyter PyYAML matplotlib

# Install R packages
RUN R -e "install.packages(c('rmarkdown', 'knitr', 'tidyverse', 'quarto'))"
```

Is this uncommon? Seems like this is basically what people who use the VSCode extension want.

**Reply by jcrben** [reply]
@mcanouil while I can render out a preview into a browser, the Visual Editor doesn't work:

```
2025-04-15 22:23:39.460 [info] node:internal/modules/esm/resolve:265
    throw new ERR_MODULE_NOT_FOUND(
          ^

Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/opt/quarto/share/editor/tools/vs-code.mjs' imported from /var/home/ben/.vscode/extensions/quarto.quarto-1.120.0/out/lsp/lsp.js
    at finalizeResolution (node:internal/modules/esm/resolve:265:11)
    at moduleResolve (node:internal/modules/esm/resolve:940:10)
    at defaultResolve (node:internal/modules/esm/resolve:1176:11)
    at ModuleLoader.defaultResolve (node:internal/modules/esm/loader:542:12)
    at ModuleLoader.resolve (node:internal/modules/esm/loader:510:25)
    at ModuleLoader.getModuleJob (node:internal/modules/esm/loader:239:38)
    at ModuleLoader.import (node:internal/modules/esm/loader:472:34)
    at defaultImportModuleDynamicallyForScript (node:internal/modules/esm/utils:227:31)
    at importModuleDynamicallyCallback (node:internal/modules/esm/utils:249:12)
    at /var/home/ben/.vscode/extensions/quarto.quarto-1.120.0/out/lsp/lsp.js:374:75886 {
  code: 'ERR_MODULE_NOT_FOUND',
  url: 'file:///opt/quarto/share/editor/tools/vs-code.mjs'
}

Node.js v20.18.2

2025-04-15 22:23:39.470 [info] [Info  - 10:23:39 PM] Connection to server got closed. Server will restart.
2025-04-15 22:23:39.470 [info] true
2025-04-15 22:23:39.801 [info] Unable to find specified quarto executable: '/opt/quarto/bin/quarto'
```

![image](https://github.com/user-attachments/assets/c3411296-7a33-4936-9151-c6ecbac8f1d3)

Should I create a new issue?

I found where it's looking for /opt/quarto/bin https://github.com/quarto-dev/quarto/blob/4e6585878fb76cbef80240ff1bf2a4a109008b89/packages/quarto-core/src/context.ts#L216

I did configure the Quarto: Path

![image](https://github.com/user-attachments/assets/cc7190e9-dd70-433f-9edf-78954e5c6e52)

**Reply by mcanouil** [reply]
> Should I create a new issue?

No, especially if you are using a one year old Quarto CLI.
Quarto is not responsible of your Docker image.
If you don't add Quarto CLI to the path, then there is no reason for tools to know where it is located.

**Reply by mcanouil** [reply]
To be clear, with a properly setup environment, there are no issues.
I'm using it with Docker images for years and more recently on GitHub Codespaces.

**Reply by jcrben** [reply]
@mcanouil where is the documentation on how to set up the environment for the VSCode extension to work? https://quarto.org/docs/tools/vscode.html contains docs on how to use it and https://marketplace.visualstudio.com/items?itemName=quarto.quarto has a short installation section saying to simply install it, but that didn't work for me

cc: @jjallaire

**Reply by mcanouil** [reply]
Your issue is unrelated to the VSCode extension.
It's your Docker image.
`quarto` needs to be available in the `PATH` for the extension to detect it. That's written in the screenshot you shared.
You are the one installing Quarto.
(When you set the path to Quarto in VSCode an extension or window reload might be needed)

Also, please refrain to ping people, especially J.J.

**Reply by jcrben** [reply]
@mcanouil quarto is available in the PATH, otherwise the preview wouldn't be working in VSCode, which as it does. Thanks for the note on the quarto-cli version, which I fixed, altho unsurprisingly it didn't affect the Visual Edit issue noted above

I pinged JJ because I noticed that he personally handled most of the issues reported over at https://github.com/quarto-dev/quarto-vscode/issues?q=is%3Aissue%20state%3Aclosed before the repo was moved to https://github.com/quarto-dev/quarto/tree/main/apps/vscode

Why is it a bad idea to ping people? It looks like you don't work for Posit so I'm guessing you don't speak to their policies.

**Reply by mcanouil** [reply]
I can't reproduce your issue using Docker images and VSCode (desktop or web).
So the issue is somewhere in your setup. Like something not up to date, etc.

When you ping people, it triggers prominent notifications. (Like pinging me in almost all your post for an unknown reason)
J.J. no longer work on this project. It's not because you find a commit by someone that you should ping that person.
Do note that your other assumption is wrong.

**Reply by mcanouil** [reply]
Try using this: https://github.com/mcanouil/quarto-codespaces/pkgs/container/quarto-codespaces

Connect to it using the `vscode` user not root.
